### DISCUSSION TITLE
Built-in accessibility checker for Wagtail

### DISCUSSION THREAD

**Discussion by thibaudcolas** [wagtail/wagtail#9262] [Ideas]
### Is your proposal related to a problem?

As part of our goal to conform with [ATAG 2.0](https://www.w3.org/TR/ATAG20/), we need a built-in accessibility checker for Wagtail. We’ve currently settled on [Sa11y](https://sa11y.netlify.app/) as a good candidate. It’s based on Tota11y, similarly to the popular [wagtail-accessibility](https://github.com/neon-jungle/wagtail-accessibility) package, but is purpose-built for content management use cases.

### Describe the solution you'd like

We should integrate Sa11y as part of Wagtail’s full page previews, and within the page editor, either in the live preview panel, or in a separate panel alongside it.

### Describe alternatives you've considered

We’ve considered two kinds of alternatives:

- Using another checker, for example Axe. Axe is very capable, but not geared specifically towards content creation, and doesn’t have a built-in UI to display its results. So it doesn’t feel like as straightforward of a fit.
- Keeping this as a third-party plugin. This seems sensible for this kind of opinionated integration, but if we do want to meet ATAG 2.0 conformance, it has to be a feature available to all CMS users (so we conform with ATAG 2.0 with the _default_ setup of the CMS, rather than _if_ the CMS is set up in a specific way

**Comment by lb-** [comment]
A few thoughts

* sa11y looks great and agree it seems well purposed to help with CMS previews
* If this is not going to be a third party package - will we provide some kind of hook (or other) to use a different library/or disable this? I think it is best to be enabled by default but understand that some developers will want to turn this off  or replace this with their own.
* will we allow for the integration of custom checks? https://sa11y.netlify.app/developers/custom-checks/
* In the preview we might be best to integrate this with the userbar (bird button) maybe showing some kind of error number on it and then when you click the bird you can click a sub-menu item to show the a11y sidebar.
* We need to be careful integrating more JS/CSS on top of any previewed page, it opens up more risk for overlapping with the page's styles or JS with naming clashes.

**Comment by Temidayo32** [comment]
I think this a great idea. Sa11y is designed for content authors and focuses on content related issues and successes. Just make sense given the target audience. It will make it easier implementing this since no too much coding is required

**Comment by thibaudcolas** [comment]
We’re starting work on this today, as an [Outreachy open source internship](https://wagtail.org/blog/outreachy-welcoming-new-contributors-to-open-source/). @albinazs is [our new contributor](https://wagtail.org/blog/our-very-first-outreachy-interns/), supported by @jams2, @thibaudcolas, @laymonage.

From now on we’ll be using this Discussions thread to share our notes and updates – starting today, for the next 13 weeks.

**Comment by thibaudcolas** [comment]
## Week 1 sprint planning

- Sprint ceremonies: planning, standups, demo & retrospective
- Backlog: see [Outreachy: Accessibility checker](https://github.com/orgs/wagtail/projects/19/views/5)
- Q&A
  - Why do we want to do this rather than keep using wagtail-accessibility? Because we want something built into Wagtail so we can turn it on by default, and if we do so we need something that’s more up-to-date, more configurable, higher-quality.
  - Is Salesforce’s `sa11y` relevant? No, same name but completely different package
  - How exactly will we configure Sa11y? Major task for us to determine – we’ll do a review first.

**Comment by thibaudcolas** [comment]
Quick mockup of our goal for the first integration approach – integrating Sa11y with the userbar:

![proposed-sa11y-poc](https://user-images.githubusercontent.com/877585/205656948-cbc6c4b5-588b-48b2-b925-a116e0ca24eb.png)

And when there are errors:

![proposed-sa11y-poc-errors](https://user-images.githubusercontent.com/877585/205657043-73bd4962-fce3-468a-9678-2facaad695c3.png)

**Comment by thibaudcolas** [comment]
## Week 2 update

- Shared details of the project with Wagtail’s accessibility team last week on Friday
- We’ve had a late planning session, with tasks visible on [our shared board](https://github.com/orgs/wagtail/projects/19/views/1)
- Albina has produced a [specifications doc](https://docs.google.com/document/d/1z4z-CmX_5H2eql8heKxK3yM7FBrjta1qM_BhLKhxl_Q/edit#) with all possible Sa11y options we could configure
- We’re discussing potential changes to Sa11y with its maintainers over at https://github.com/ryersondmp/sa11y/issues/41

Last but not least, @albinazs wrote about her experience so far! [How I got my first internship as a junior front-end developer
](https://dev.to/albinazs/how-i-got-my-first-internship-as-a-junior-front-end-developer-38h5)

**Comment by albinazs** [comment]
### **Mid-internship update**
As Sa11y changed their license to GPL, we’ve decided it’s enough of a concern for us that we won’t integrate Sa11y directly in Wagtail. We created a custom accessibility checker based on the [axe-core](https://github.com/dequelabs/axe-core) testing engine with UI designed for content authors to easily identify and fix accessibility issues. The first integration is within the userbar - the corresponding PRs were merged last week, meaning that the feature will be included in the Wagtail 4.2 release.

The checker scans the loaded page for errors and displays the results in the GUI, with 3 initial rules (`empty-heading`, `p-as-heading`, `heading-order`). It can be turned off with the `construct_wagtail_userbar` hook if needed.

![image](https://user-images.githubusercontent.com/51043550/214340572-1a949abd-0fd1-4478-82c7-2922840a6dae.png)

We are planning to improve the checker and following a discussion with the accessibility team last week, our priorities include:
- Expanding the number of rules that the checker runs in addition to the initial three
- Implementing an outline feature that highlights inaccessible elements on a page
- Live edit panel integration

The full list of ideas can be checked in the meeting minutes [here.](https://docs.google.com/document/d/1YUxOs5jYZMd8rX291mDE123xIK7tbD63PzSR9ooFa4c/edit?usp=sharing)

We would love to hear any feedback and suggestions for further improvements.

**Comment by thibaudcolas** [comment]
Quick update based on recent work – and because we’re almost reaching the end of @albinazs’s internship 🥲.

This happened:

- Wrote a [brief overview of how the checker works](https://docs.wagtail.org/en/latest/advanced_topics/accessibility_considerations.html#built-in-accessibility-checker) in the Wagtail documentation. This is intended for developers, we will likely produce similar material for CMS users in the future.
- Added 5 new rules to the checker – bringing the current total to 8 (see the previous link to know which rules are present)
- Got feedback from lots more folks – name dropping in particular people at [Syracuse University](https://www.syracuse.edu/), [RNIB](https://www.rnib.org.uk/), and [NASA JPL](https://www.jpl.nasa.gov/).

Here are remaining improvements we intend to do during the internship or shortly after:

- [Sorting accessibility checker results based on their order on the page #10013](https://github.com/wagtail/wagtail/issues/10013)
- [Highlight elements with errors in accessibility checker (outline) #10138](https://github.com/wagtail/wagtail/issues/10138)
- [Keyboard support improvements in accessibility checker #10135](https://github.com/wagtail/wagtail/issues/10135)

Things that are on our radar in the near future (hopefully in time for the next release):

- [Integrate accessibility checker within the page editor #10136](https://github.com/wagtail/wagtail/issues/10136)
- [Support custom configuration options in accessibility checker #10137](https://github.com/wagtail/wagtail/issues/10137)

Our full [project backlog](https://github.com/orgs/wagtail/projects/19/views/1) showcases other improvements which we’d like to do but we are unlikely to get to at this time.
